Deploy MSI build artifacts

Boddlnagg opened the pull request:
This enables deployment of `rustup.msi` and the corresponding SHA of the embedded executable. I don't know exactly how the corresponding AppVeyor settings work, so whoever is responsible for the original setup should have a look at this, since I'm not able to test it locally.

brson commented:
The windows failure looks legit. Not sure what's going on with the unix builds, but it looks similar to something I saw building an older cargo rev recently. It *looks* like the hash of the openssl tarball changed!

I suspect the appveyor settings for 'artifacts' and 'deployment' aren't quite right but I'll have to experiment with it myself.

Thanks @Boddlnagg.

Boddlnagg commented:
I'm not sure if this will actually work, since the artifacts `rustup-msi` and `rustup-msi-exe-sha` will only exist for the `BUILD_MSI` configuration. But maybe AppVeyor does the right thing and just ignores the artifacts if the files in `dist` don't exist (it looks like that's what's happening for the cases where `prepare-deploy-appveyor.ps1` just exits without putting anything in `dist`).

brson commented:
The appveyor failure is unrelated to this pr. Looks like a race in a pre-existing test. Trying to figure it out now.

brson commented:
I've also fixed the travis failures in a separate PR.

Boddlnagg commented:
I just rebased this and removed some unrelated changes that belong to another PR. I have more changes ready, but I thing I'm going to wait until this is merged.

brson commented:
Thanks @Boddlnagg. Fighting fires now but haven't forgotten.

bors commented:
:umbrella: The latest upstream changes (presumably #776) made this pull request unmergeable. Please resolve the merge conflicts.

Diggsey commented:
@brson We need to do something with this - can I do anything to help?

brson commented:
@Diggsey Yeah we do.

I'm afraid I haven't looked too closely at the solution here, so I'm not sure the state of it.

If you wanted you could try to get it building again and seeing if you can get it to pass the CI.

Boddlnagg commented:
I can offer some help if needed, though my time is limited.

> I have more changes ready, but I think I'm going to wait until this is merged.
According to this earlier comment of mine, I also have some more changes that build upon this PR ... I am going to see what that was and can try to prepare a PR.

brson commented:
@Boddlnagg thanks for the updates. Sorry for avoiding this for so long. I left a comment on the msi issue in response.

alexcrichton commented:
Closing in favor of https://github.com/rust-lang-nursery/rustup.rs/pull/1211 which looks like it's got some more action
